bpo-27645: Supporting native backup facility of SQLite

lelit opened the pull request:
This is a new version of PR #377, rebased on more recent `master` to solve the conflicts on the old `NEWS` file. 

It implements a [new requested feature](https://bugs.python.org/issue27645#msg304789) since last review.

See https://bugs.python.org/issue27645

<!-- issue-number: bpo-27645 -->
https://bugs.python.org/issue27645
<!-- /issue-number -->

lelit committed:
Expose the SQLite Online Backup API as Connection.backup()

See issue #27645.

lelit committed:
Preliminary documentation for sqlite3.Connection.backup()

lelit committed:
Reduce code nesting depth by using a finally label, as other functions do

lelit committed:
Wrap SQLite calls between BEGIN_ALLOW_THREADS and END_ALLOW_THREADS

Suggested by Aviv Palivoda.

lelit committed:
Make pages and progress keyword-only args in sqlite3.Connection.backup()

lelit committed:
Remove reference to SQLite docs in Connection.backup() documentation

lelit committed:
Use f-string in the sqlite3.Connection.backup() example

lelit committed:
Explicitly mark sqlite3.Connection.backup() as added in v3.7

lelit committed:
Reduce chances of cluttering future “git blame”

lelit committed:
Parametrize the name of the database involved in the backup

Suggested by Aviv Palivoda.

lelit committed:
Test propagation of exception raised in backup's progress callback

lelit committed:
Use a better name for variable

lelit committed:
Do not delay next iteration if the result was OK

When sqlite3_backup_step() returns OK it means that it was able to do it's work but there are
further pages to be copied: in such case it's better to immediately start the next iteration,
without opening a window that allows other threads to go on and possibly locking the database.

Suggested by Aviv Palivoda.

lelit committed:
Omit the backup method when underlying SQLite library is older than 3.6.11

lelit committed:
Assert that the non-mandatory arguments are keyword-only

lelit committed:
Pass also the current status of the ongoing backup to the progress callback

Add new SQLITE_DONE integer constant to the module, that could be used by the callback to
determine whether the backup is terminated.

lelit committed:
Slightly different way handling backup step's error state

This should address Aviv's concerns about how the sqlite3_backup_step() errors are handled.

lelit committed:
When an error occurs while the backup is going on, remove the target file

lelit committed:
Add NEWS entry for issue 27645

lelit committed:
New argument "sleep" to specify a different number of milliseconds

lelit committed:
Allow to copy the database to either an external file or to another Connection

lelit committed:
Check that target backup DB is not in transaction

This is needed for SQLite < 3.8.7.

lelit commented:
Compilation [fails](https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.7951#L498) on Windows, due to misdeclaration of `unlink()`. I call that function to remove an incomplete backup when an error happens: I looked around in the sources and couldn't figure out a portable way to do that. Any hint?

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -21,6 +21,12 @@
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#else
+extern int unlink(const char *);
```
This won't work under Windows. You have to use ``DeleteFileW``, see posixmodule.c

lelit committed:
Use DeleteFileW() under MS Windows to delete incomplete backup

lelit committed:
Second attempt at properly calling DeleteFileW()

Follow Christian Heimes suggestions to fix DeleteFileW() declaration and usage.

lelit committed:
Setup exception from sqlite error condition

lelit committed:
Close the backup connection in case of error

lelit committed:
Attempt to chain the exception raised by incomplete backup deletion

lelit committed:
Fix double call to PyMem_Free()

lelit committed:
Remove leftover comment

lelit committed:
Drop const qualifier to avoid compiler warning

lelit committed:
Avoid using NamedTemporaryFile in test cases

MS-Windows does not allow to operate on a file while it's being used by some process: let's see
if this fixes the CI errors on AppVeyor.

tiran commented on Lib/sqlite3/test/backup.py:
```diff
@@ -0,0 +1,166 @@
+        self.assertEqual(result[0][0], 3)
+        self.assertEqual(result[1][0], 4)
+
+    def CheckBadTarget(self):
```
Please don't follow the example of the other sqlite tests and use standard test names ``test_foo``.

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,173 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    PyObject* target = NULL;
+    const char* filename;
+    int pages = -1;
+    PyObject* progress = Py_None;
```
PEP 7, the ``*`` come before the var name, not after the type. It's ``PyObject *progress``. Same for ``char *name``.

lelit replied:
I've used the same style as the whole unit, do you think I should follow PEP7 in only new function?
I personally prefer that style, but seemed better to follow the code around... the same reasoning I had with the backup.py (but that's completely new...)

tiran replied:
Are you sure? :)
```
+    sqlite3 *bckconn;
+    sqlite3_backup *bckhandle;
```
I guess muscle memory is hard to beat...

Your remark is valid. I don't care that much about PEP 7, as long as you stick to one one style. It's really up to you, either obey PEP 7 for new code or stick to the existing style.

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,173 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        }
+    }
+
+finally:
```
You have defined a finally target but don't use it to clean up resources. Traditionally Python's C either uses goto to perform cleanups or returns immediately.

lelit replied:
Ok.

tiran commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,54 @@ Connection Objects
+      database, ``"temp"`` to indicate the temporary database or the name specified
+      after the ``AS`` keyword in an ``ATTACH`` statement for an attached database.
+
+      The *sleep* argument specifies the number of milliseconds to sleep by between
```
Almost all CPython functions and methods use seconds as double for timeout. Milliseconds is going to confuse people.

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,173 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    sqlite3_backup *bckhandle;
+    static char *keywords[] = {"target", "pages", "progress", "name", "sleep", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|$iOsi:backup", keywords,
```
See other comment, Format ``d`` accepts ints and floats and converts them to double. Multiple the double with 1000 and cast it to int ``sqlite3_sleep(sleep)``

lelit replied:
Ok, will do!

lelit committed:
Make the sleep parameter a double, accepting a value in seconds

lelit committed:
Replace gotos with explicit return, given that no cleanup is needed

lelit committed:
Rename test methods using snake_style instead of CamelCase

lelit committed:
Use one single pointers declaration style in the new function

The source unit already uses a mix of styles (4-spaces vs 8-spaces indents, "char* name" vs
"char *name") so it's difficult to select "the right one": stick with PEP7.

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        return NULL;
+    }
+
+    if (progress != Py_None && !PyCallable_Check(progress)) {
```
Sorry, I missed this. We prefer ``PyNone_Check()`` these days.

serhiy-storchaka replied:
`PyNone_Check()` doesn't exist.

lelit replied:
Uhm, really? The only reference I can find is in `Doc/c-api/none.rst` and it says:
```
Note that the :c:type:`PyTypeObject` for ``None`` is not directly exposed in the
Python/C API.  Since ``None`` is a singleton, testing for object identity (using
``==`` in C) is sufficient. There is no :c:func:`PyNone_Check` function for the
same reason.
```
Am I wrong?

tiran replied:
Nope, you are right and I got fooled by some legacy code base. I have been banging my head against old C code for the last weeks. The code base has ``#define PyNone_Check(o) ((o) == Py_None)``.

Scratch my comment.

tiran commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+            PyErr_SetString(PyExc_ValueError, "target cannot be the same connection instance");
+            return NULL;
+        }
+#if SQLITE_VERSION_NUMBER < 3008007
```
3.8.7 was released in 2014, https://www.sqlite.org/releaselog/3_8_7.html

Do we need to support older versions? How about requiring a minimum version of 3.8.7 or even newer to compile CPython?

benjaminp replied:
I'm definitely in favor of not supporting old versions of dependencies if it lets us avoid lots of compatibility code.

tiran replied:
According to distrowatch, Debian Jessie has 3.8.7.1 but CentOS 7-1708 only 3.7.17. Guess we need to support < 3.8 for a while.

tiran reviewed (changes requested):
You have touched the sqlite C code in the last year. Please have a lot, especially at the Windows code.
@berkerpeksag @benjaminp @serhiy-storchaka 

What do you think about requiring a sensible modern version of sqlite to compile the extension module in 3.7?

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    const char *filename;
+    int pages = -1;
+    PyObject *progress = Py_None;
+    char *name = "main";
```
Why not `const char *`?

lelit replied:
Right, fixed.

serhiy-storchaka commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,55 @@ Connection Objects
+
+      By default, or when *pages* is either ``0`` or a negative integer, the entire
+      database is copied in a single step; otherwise the method performs a loop
+      copying up to the specified *pages* at a time.
```
I think `*` around `pages` is not needed here.

lelit replied:
You mean only the first reference to the argument should be emphasized?

serhiy-storchaka replied:
I think this isn't a reference to the argument, but a unit. If you want to use a reference to the argument you should write
```
copying up to *pages* pages at a time
```

lelit replied:
Ok.

serhiy-storchaka commented on Doc/library/sqlite3.rst:
```diff
@@ -521,6 +521,55 @@ Connection Objects
+         import sqlite3
+
+         source = sqlite3.connect('existing_db.db')
+         dest = sqlite3.connect(':memory')
```
`':memory:'`

lelit replied:
Fixed.

serhiy-storchaka commented on Lib/sqlite3/test/backup.py:
```diff
@@ -0,0 +1,166 @@
+    runner.run(suite())
+
+if __name__ == "__main__":
+    test()
```
Does not `unittest.main()` work?

lelit replied:
Yes, now that the backup suite uses _snake_like_ method names. Done.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2017-10-05-20-41-48.bpo-27645.1Y_Wag.rst:
```diff
@@ -0,0 +1,2 @@
+sqlite3.Connection now exposes a backup() method, if the underlying SQLite
+library is at version 3.6.11 or higher. Patch by Lele Gaifax
```
Period.

lelit replied:
Done.

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        pages = -1;
+    }
+
+    if (PyUnicode_Check(target)) {
```
Is it worth to support path-like objects?

lelit replied:
Why not indeed... but at that point, should I also change the other functions accepting a path name?

serhiy-storchaka replied:
On other hand, it may be better to remove support of a file name. The code will be simpler and more reliable if accept only a Connection instance.

lelit replied:
Uhm, yes, _slightly_ simpler indeed, but that would shift the complication to the caller site: in all my use cases I'd use the file name form, so if that gets dropped I'd need to set up another connection instead...
I added the connection form to deal with Cédric's _opposite_ use case, to _restore_ the content of a test database.
I'm -0 on removing such support, but leave the final decision to you.

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        rc = sqlite3_open(filename, &bckconn);
+        Py_END_ALLOW_THREADS
+
+        if (rc != SQLITE_OK) {
```
`pysqlite_connection_init()` uses a different code for handling errors. Is there an error in `pysqlite_connection_init()` or the code here can be simpler?

lelit replied:
From my understanding of the [first paragraph](http://www.sqlite.org/c3ref/open.html), I would say yes: `sqlite3_open()` always return a new database connection, with only one exception, OOM.
Do you agree?

If yes, what's the best approach:
a. fix `pysqlite_connection_init()` now
b. defer the fix to a different PR
?

serhiy-storchaka replied:
I think a separate issue should be open for fixing `pysqlite_connection_init()`. Since this is a bug fix, it needs to be backported to other branches.

lelit replied:
See [#32274](https://bugs.python.org/issue32274): I was wrong about that, the database connection will be released by the `pysqlite_connection_dealloc()` function in any case.

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+            Py_END_ALLOW_THREADS
+
+            if (progress != Py_None) {
+                if (!PyObject_CallFunction(progress, "iii", rc,
```
Leaks the result.

lelit replied:
Good catch, fixed!

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+    bckhandle = sqlite3_backup_init(bckconn, "main", self->db, name);
```
Is it worth to make the destination table name configurable?

lelit replied:
The second parameter, `"main"`, is the destination _database_ name (see [here](http://www.sqlite.org/c3ref/backup_finish.html#sqlite3backupinit)): in principle yes, it could be exposed as a parameter, but while the source name (ie the fourth parameter) is useful because you may desire to take a snapshot of an _attached_ database, using a target different from `"main"` seems seldom useful...

lelit committed:
Fix glitch in example code

lelit committed:
Rectify explanation of the pages parameter

lelit committed:
Add missing ending dot

lelit committed:
Simplify code

No need to setup a special TestSuite, now that the test method names are in the "snake style".

lelit committed:
Add missing const qualifier

lelit committed:
Fix leak by DECREFing the result of progress callback call

lelit committed:
Restore the suite() function, used by outer Lib/test/test_sqlite.py

lelit committed:
Simplify backup() implementation

Now the target argument must be exclusively another sqlite3 connection, to avoid dealing with
filenames directly from C.

berkerpeksag commented on Modules/_sqlite/connection.c:
```diff
@@ -21,6 +21,17 @@
+#else
+#ifdef MS_WINDOWS
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
```
Do we still need this after a2f15bc?

berkerpeksag commented on Modules/_sqlite/connection.c:
```diff
@@ -1442,6 +1457,122 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+        if (rc == SQLITE_NOMEM) {
+            (void)PyErr_NoMemory();
+        } else {
+            PyErr_SetString(pysqlite_OperationalError, sqlite3_errstr(rc));
```
Unfortunately, `sqlite3_errstr` was added in 3.7.15. So we need either add fallback or require SQLite 3.7.15 instead of 3.6.11.

berkerpeksag commented on Modules/_sqlite/connection.c:
```diff
@@ -21,6 +21,17 @@
  * 3. This notice may not be removed or altered from any source distribution.
  */
 
+#include "Python.h"
```
It looks like all these header additions can be removed after a2f15bc.

lelit committed:
Remove now useless includes

serhiy-storchaka commented on Modules/_sqlite/connection.c:
```diff
@@ -1478,9 +1477,27 @@ pysqlite_connection_backup(pysqlite_Connection *self, PyObject *args, PyObject *
     }
 
-    if (!PyUnicode_Check(target) && !PyObject_TypeCheck(target, &pysqlite_ConnectionType)) {
+    if (!PyObject_TypeCheck(target, &pysqlite_ConnectionType)) {
```
You can use the `O!` format unit in `PyArg_ParseTupleAndKeywords()`.

lelit committed:
Recognize common error codes for old SQLite3 missing sqlite3_errstr()

lelit committed:
Move argument type check to PyArg_ParseTupleAndKeywords()

lelit committed:
Use better names for local variables

berkerpeksag committed:
Merge remote-tracking branch 'upstream/master' into pr/4238

berkerpeksag committed:
Cosmetic fixes

berkerpeksag committed:
Fix segfault in test_bad_target_closed_connection

berkerpeksag committed:
Relax an assert

berkerpeksag committed:
This will be included in 3.7
